pyccel/parser/semantic.py

```diff
@@ -2454,20 +2546,15 @@ def _visit_FunctionDef(self, expr, **settings):
                     local_vars += [var]
 
             if 'stack_array' in decorators:
-
                 for var in local_vars:
-                    var_name = var.name
-                    if var_name in decorators['stack_array']:
+                    if var.name in decorators['stack_array']:
                         var.is_stack_array = True
                         var.allocatable    = False
                         var.is_pointer     = False
```

Comment:
In this case we are visiting a `FunctionDef`, so we have access to the decorators directly through `expr.decorators`. We could get them from the Scope, but I see no reason for that.

Reply:
As soon as you visit the body of the `FunctionDef` you don't have access to `expr.decorators` anymore. That's where we use the Scope.

Reply:
But this is still being done after the body has been printed, so after all the variables have been created. If it was earlier and just used for arguments then it would make sense, but here the arguments are not set until after they are used, and the other variables are set twice. Or have I missed something?

Reply:
Sorry, I see what you mean now! You are probably right.

I suppose we should do this only for the function arguments, and before we visit the function body.

Reply:
There are a few things that are not clear to me in `_visit_FunctionDef`, so I cannot say for certain if this will work.
@saidctb What do you think?

Reply:
yes we should delete this part if it's already been done when the variable was ceated

Reply:
Actually this is only done for local variables, and not for function arguments or results:
```python
            if 'stack_array' in decorators:
                for var in local_vars:
                    if var.name in decorators['stack_array']:
                        var.is_stack_array = True
                        var.allocatable    = False
                        var.is_pointer     = False

            if 'allow_negative_index' in decorators:
                for var in local_vars:
                    if var.name in decorators['allow_negative_index']:
                        var.allows_negative_indexes = True
```
Hence it looks like these lines could be simply removed. But it remains to be determined what we should do with the function arguments themselves, do you guys know where in the code we process their decorators?

Reply:
I've removed this lines in commit 1921f2f